crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -0,0 +1,406 @@
+use std::{alloc::Layout, cell::UnsafeCell, marker::PhantomData, ptr::NonNull};
+
+#[derive(Debug)]
+pub struct BlobVec {
+    item_layout: Layout,
+    capacity: usize,
+    len: usize,
+    data: UnsafeCell<NonNull<u8>>,
+    swap_scratch: UnsafeCell<NonNull<u8>>,
+    drop: unsafe fn(*mut u8),
+}
+
+impl BlobVec {
+    pub fn new(item_layout: Layout, drop: unsafe fn(*mut u8), capacity: usize) -> BlobVec {
+        unsafe {
+            let mut blob_vec = BlobVec {
+                swap_scratch: UnsafeCell::new(NonNull::dangling()),
```

Comment:
```suggestion
                swap_scratch: if item_layout.size() == 0 {
                    UnsafeCell::new(NonNull::dangling()),
                } else {
                    blob_vec.swap_scratch =
                        UnsafeCell::new(NonNull::new(std::alloc::alloc(item_layout)).unwrap());
                }
```
Just in case there is a panic before `blob_vec.swap_scratch` is set this would prevent trying to deallocate a dangling pointer in the `BlobVec` drop impl. I don't think a panic is possible in between, but better safe than sorry.